### ISSUE TITLE
[NEXT-1308] Css is imported multiple times and out of order in /app dir

### ISSUE THREAD

**Issue by ssijak** [vercel/next.js#51030]
### Verify canary release

* [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Mon Apr 24 20:52:24 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.12.1
      npm: 8.19.2
      Yarn: 1.22.19
      pnpm: 7.18.1
    Relevant packages:
      next: 13.4.5-canary.9
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue or a replay of the bug

[https://github.com/ssijak/next-css-issue-not-working-simple](https://github.com/ssijak/next-css-issue-not-working-simple)

### To Reproduce

Just start the app and check the styling on the buttons. Styles are imported multiple times wherever `Button` was used (page and layout) and order is also not deterministic, so it can be imported in different order on different app runs.

This is another/same simple repro difference is just that it uses turbo and transpiles the UI lib, I started with that but figured that the issue is happening without it too [https://github.com/ssijak/next-css-issue-not-working](https://github.com/ssijak/next-css-issue-not-working)

### Describe the Bug

\-Same styles are imported multiple times
\-Order of imports is not deterministic

Screenshot: [https://share.cleanshot.com/nq35j7vh](https://share.cleanshot.com/nq35j7vh)

![](https://uploads.linear.app/099ed59a-06c6-472f-9977-4798721bda90/1fe54a6f-2c7b-45a7-bcbf-3b4fcd6580af/ab619c5d-aa73-4efb-973f-0d0d21cf8f34)

### Expected Behavior

Same styles should be imported once. Starting the app multiple times should not produce different results (ordering of CSS, impacting specificity)

### Which browser are you using? (if relevant)

*No response*

### How are you deploying your application? (if relevant)

*No response*

<sub>From [SyncLinear.com](https://synclinear.com) | [NEXT-1308](https://linear.app/vercel/issue/NEXT-1308/css-is-imported-multiple-times-and-out-of-order-in-app-dir)</sub>

**Comment by ssijak** [comment]
This PR https://github.com/vercel/next.js/pull/50406 seems like it should have fixed this issue, but I see the problem in my repro repositories

**Comment by ssijak** [comment]
Css modules seem to be working ok. Removed vanilla-extract from this repro, and imported the same CSS module in page and layout and set it on the `<button>` and it imports the CSS once.

**Comment by igordanchenko** [comment]
This issue is still present in the latest canary (13.4.7-canary.1)

1) Same styles are imported multiple times
2) The order of imports is not deterministic

Here is a minimal repro with easy-to-reproduce test cases.

https://codesandbox.io/p/sandbox/next-51030-hlj722?file=%2Fapp%2Fpage.tsx%3A1%2C1

**Comment by tinleym** [comment]
The indeterministic order of imports wreaks havoc when CSS Cascade Layers are being used. The defined layer precedence that belongs in the root layout file gets read after the application of layers in child components. This causes the defined layer precedence to be ignored in favor of a more or less random precedence, and can make routes unusable. 

This issue surfaced when updating up from 13.4.4 to either 13.4.5 or 13.4.6.

**Comment by ddoan** [comment]
One workaround I'm currently using is to make the children CSS *more* specific: `button.button` instead of `.button` or `div.something` instead of `.something`.

**Comment by shuding** [comment]
For https://codesandbox.io/p/sandbox/next-51030-hlj722?file=%2Fapp%2Fpage.tsx%3A1%2C1, it becomes very tricky if your CSS code itself has conflicts and some behaviors are undefined. For example when a dynamic component has loaded, its CSS might cause side effects to other parts of the page, which is expected.

For example, if you have a layout like this:

```js
import style1 from 'a.module.css'
import style2 from 'b.module.css'

<button className={style1.btn + ' ' + style2.btn}>
```

The styles that imports later will override previous ones, so `b.module.css` will take precedence. However, if we have a page component that imports `a.module.css` again, it will take precedence over `b.module.css` again because page's styles have higher priority than the layout's. This is also a cause of duplicated styles.

In general, we recommend you to use CSS `@layer`, or try to make CSS selectors as pure and specific as possible.

**Comment by igordanchenko** [comment]
@shuding, I appreciate you taking the time to look into this example.

> it becomes very tricky if your CSS code itself has conflicts and some behaviors are undefined.

I wouldn't call styles override a conflict. Overriding base component styles in a descendant component is a pretty common technique. Isn't that what "C" in CSS stands for?

> For example when a dynamic component has loaded, its CSS might cause side effects to other parts of the page, which is expected.

The above example is a module-scoped CSS that doesn't have any global side-effects.

> The styles that imports later will override previous ones, so b.module.css will take precedence. 

Yes, you are absolutely correct, and that's the desired behavior that we are unfortunately not able to achieve under the app router.

> However, if we have a page component that imports a.module.css again, it will take precedence over b.module.css again because page's styles have higher priority than the layout's. This is also a cause of duplicated styles.

In theory, yes. But it's not the case in the above example, isn't it?

**Comment by tinleym** [comment]
@shuding if you'll please read my comment below, @layer is not working.

> The indeterministic order of imports wreaks havoc when CSS Cascade Layers are being used. The defined layer precedence that belongs in the root layout file gets read after the application of layers in child components. This causes the defined layer precedence to be ignored in favor of something random (https://css-tricks.com/css-cascade-layers/#aa-establishing-a-layer-order), and can make routes unusable.
> 
> This issue surfaced for me when updating up from 13.4.4 to 13.4.5 or 13.4.6.

**Comment by shuding** [comment]
Thanks! I'll go through all the reproductions and think about an improvement in the next couple of days!

I understand that the confusion caused by the new architecture. In general scoped CSS itself doesn't have conflicts, but the order and duplication might cause the styles to conflict (override each other). I'll give you an example:

layout: <button class='base btn'> → layout.css
page: <button class='base'> → page.css

If the module that contains the `.btn` class is imported lastly, it will override the styles defined in `.base`. If **only** layout is rendered, that's the expected behavior.

However when page is rendered, `page.css` will be loaded and it has only `.base`. This CSS file will occur after `layout.css` and override the `.btn` styles in that file.

Previously in the pages directory, all styles will be packed into one CSS file. But here we need to generate one CSS per layout and page because they can be loaded separately.

That said, we still have some ideas to improve it. Like changing the CSS modules hashing algorithm to make it base on the layer, or deduplicate styles that are already in its parent layer. Will keep you updated in this issue.

**Comment by boatilus** [comment]
> That said, we still have some ideas to improve it. Like changing the CSS modules hashing algorithm to make it base on the layer, **or deduplicate styles that are already in its parent layer.**

If I understand the latter suggestion correctly, this seems to me like the right approach. If a build system produces hashed rule/class names based on the rule contents (e.g., `._129ac00b`) in `layout.css`, it'd be unnecessary to duplicate that rule in `page.css`, and that duplication causes the unexpected precedence fighting. The basic idea here would be `if (selector in layout.css) strip from page.css`. Although I say that with that big caveat that I don't know the implications when overriding a rule generated in `page.css` that's *intentional* and how to signify that.

To your point, if there's perfect purity, this won't be a problem, but it's also reasonably unrealistic that there wouldn't be things like rules applied to descendent selectors (and probably some non-negigible number of other scenarios) on class names that have a high likelihood of their precedence being trumped by the same properties in a rule duplicated in `page.css`.

**Comment by sleepdotexe** [comment]
I'm also encountering the same behaviour.

For my use case, I'm creating standard UI/layout components that I want to use across multiple pages and components – these components have `*.module.css` styles. This allows me to add default custom stylings to these UI components and only load them if the component is used on the page. I also allow an additional `className` to be added for additional styling at a per-component level.

The problem is, as mentioned above, changing between routes/different app runs can result in different CSS load orders in addition to a huge amount of duplicate CSS.

For example:
```jsx
// components/layout/Section.tsx

interface Props extends React.ComponentProps<'section'> {
	fullWidth?: boolean;
}

export default function Section(props: Props) {
	const { fullWidth, ...remainingProps } = props;
	return (
		<section
			{...remainingProps}
			className={[styles['section'], props.className].join(' ')}
			data-fullwidth={fullWidth}
		>
			{props.children}
		</section>
	);
}
```

Navigating from `Page A`, to `Page B`, back to `Page A` cause `Page A` to display differently the second time round. Refreshing manually causes `Page A` to revert to its original layout.

For now, as a partial fix, I have managed to use css `@layer` to lower the CSS specificity of all the UI components' styles, however it still results in large amounts of duplicate css being added to each page, especially if the UI component is used frequently (such as a custom button).

Would be interested to know if there is a better way to go about this. I can provide more detailed examples/reproductions if necessary.

**Comment by jep-a** [comment]
@tinleym 
> The current order of imports wreaks havoc when CSS Cascade Layers are being used. The defined layer precedence that belongs in the root layout file gets read after the application of layers in child components. This causes the defined layer precedence to be ignored in favor of something random (https://css-tricks.com/css-cascade-layers/#aa-establishing-a-layer-order), and can make routes unusable.
> 
> This issue surfaced for me when updating up from 13.4.4 to 13.4.5 or 13.4.6.

Have a quick hack fix for the css layer order here https://github.com/vercel/next.js/issues/16630#issuecomment-1644918877

**Comment by katiasmet-93** [comment]
Is there already a solution for css that is loaded multiple times if you use the "use client" mode? 
I use css modules and have f.e. several buttons or inputs on 1 page. The css (modules) are loaded multiple times in version 13.4.19.

**Comment by monolithed** [comment]
Is there a way to consolidate all styles into one file, so that the minifier automatically removes duplicates? I can't estimate the scale of the issue in kilobytes, but I see that each of my components (a total of 31) duplicates styles three times, and this is in production.

<img width="845" alt="Screenshot 2023-09-05 at 01 37 57" src="https://github.com/vercel/next.js/assets/803674/38166b25-5153-4c4f-8583-7ea823caf558">

**Comment by Stillonov** [comment]
I have the same problems and it's awful. My app is very fragile.
This happens when I navigate through pages containing a common component. Sometimes it looks like the cache from the last page is being saved and duplicated.

<img width="409" alt="Screenshot 2023-09-08 at 01 24 15" src="https://github.com/vercel/next.js/assets/11207734/0a5aa192-e3b6-4f9b-9d08-45f8a8631ed3">

**Comment by katiasmet-93** [comment]
Maybe not the best solution for now, but converting to the pages router fixed it for me.

**Comment by monolithed** [comment]
> Maybe not the best solution for now, but converting to the pages router fixed it for me.

@katiasmet-93, your solution may be suitable for projects with a small codebase where there is no need for investment in testing. However, such thoughts do make one consider a complete abandonment of Next. This is already the third significant bug on my record, following the poorly functioning Image tag and unnecessary requests. Does anyone really use Next in serious projects?

**Comment by nikita2090** [comment]
I have the same problem when using shared components in layouts and pages (next 13.4.19). It doesn't work correctly both when using CSS modules and when using tailwind.

**Comment by olalonde** [comment]
Same issue with mantine https://github.com/vercel/next.js/issues/16630#issuecomment-1718974809

**Comment by gffuma** [comment]
Same problem on 13.5.2 not the ideal solution but downgrade to 13.4.4 helped me.

**Comment by monolithed** [comment]
> Same problem on 13.5.2 not the ideal solution but downgrade to 13.4.4 helped me.

<img width="407" alt="Screenshot 2023-09-22 at 19 04 29" src="https://github.com/vercel/next.js/assets/803674/7964f34c-171b-4667-86e6-416dc0ae6481">


@gffuma, I tried downgrading to version 3.4, and it didn't help at all. If the component is not being reused, there are no issues.

**Comment by gffuma** [comment]
Is exact the 3.4.4 ?

**Comment by monolithed** [comment]
> Is exact the 3.4.4 ?

@dgagn, yep! I tried that version. I don't think any magic should happen because this bug has been around for more than three years

**Comment by ensidialabs** [comment]
Most likely I have the same problem. Is the normal operation of the CSS Module not available with new versions of Next.js?

![270266163-017fb489-cd28-400c-b812-967444e140e0](https://github.com/vercel/next.js/assets/105992892/dffdf896-ea2e-4f1e-b340-5f4db9fd45bb)
![270264894-b1b696ef-c3a4-46e4-b01c-2438ad7ed090](https://github.com/vercel/next.js/assets/105992892/fb399836-6dec-4baf-ae91-6ce7e5f1079a)

**Comment by manavm1990** [comment]
In my case, I can go into `layout.tsx` and just manually change the order of the CSS imports. It doesn't even matter what they are. As soon as the app hot reloads, the CSS is 'fixed,' and stays fixed even if I revert the CSS back to the original!

**Comment by tinleym** [comment]
The multiple imports also causes issue animation issues. If you want, say, a no-js fade in for a module on a dedicated route, a CSS module with an animate property could get reapplied during the animation and cause it to repeat.

**Comment by maiconsanson** [comment]
Using `14.0.1`

<img width="733" alt="Screenshot 2023-11-02 at 11 07 16" src="https://github.com/vercel/next.js/assets/1965455/131ac4b7-b8b3-4eba-9a42-1c8e96a907c3">

**Comment by tomjvdps** [comment]
For css from libraries, the only workaround I found is to **rewrite** it by encapsulating it in a `@layer`
@sjahanmard 

For defining `@layer`, I use this workaround:
https://github.com/vercel/next.js/issues/51030#issuecomment-1644922285

After all this, despite the duplication of the CSS, everything is ok
I'm using version `13.4.7`

**Comment by SkyGopnik** [comment]
Found another case of this, I'm on the `faq` page, after that I go to the `login` page and go back using browser back button, i have multiple css files on my `faq` page, that don't apply to that.

![image](https://github.com/vercel/next.js/assets/26293210/364a58c4-3eb0-4550-93c1-0fcfc8eb1687)
![image](https://github.com/vercel/next.js/assets/26293210/7a2b9aa9-3db4-46db-a88f-0c5208829321)

**Comment by pvaladez** [comment]
Global CSS is also sometimes being added after the module css(see below).  This breaks the case where you are expecting a module class to override a class from your global css file.  Even worse, the dev server seems to act differently than `next start`.  Not being able to import css in a logical and easily predictable order seems like a pretty major problem.

I've forked the codepen example from @igordanchenko and added a `global.css` that sets the button color to green:
```css
.button {
  background-color: green;
}
```

My expectation is that global CSS will **_always_** be added to the page before any components styles from css modules, so the css in `Button.module.css` will override the above `.button` selector, assuming the two selectors have the same specificity(ex. `classes.root` from `Button.module.css`).

Interestingly, the buttons on the home page will only be green if both `global.css` is the last file to be edited and its file contents have actually changed.

**So it seems to be setting the order of the CSS based on which file was edited last!?**


[https://codesandbox.io/p/devbox/next-51030-forked-hj9qmz](https://codesandbox.io/p/devbox/next-51030-forked-hj9qmz)

**Comment by AllenWilliamson** [comment]
> The current order of imports wreaks havoc when CSS Cascade Layers are being used. The defined layer precedence that belongs in the root layout file gets read after the application of layers in child components. This causes the defined layer precedence to be ignored in favor of something random (https://css-tricks.com/css-cascade-layers/#aa-establishing-a-layer-order), and can make routes unusable.
> 
> This issue surfaced for me when updating up from 13.4.4 to 13.4.5 or 13.4.6.

I solved the cascade layer issue by adding a `webpack.BannerPlugin` similar to the following:
```js
const config = {
    webpack: (
        config,
        { buildId, dev, isServer, defaultLoaders, nextRuntime, webpack }
    ) => {
        config.plugins.push(
            new webpack.BannerPlugin({
                banner: '@layer one, two, three;\n\n',
                raw: true,
                entryOnly: true,
                include: /\.css$/,
            })
        );
        return config;
    },
};

```
It's not ideal but works for now.

**Comment by AnastasiiaRoch** [comment]
The same issue is in Next 14.1.0 (SCSS Modules). The problem happens when using generic reusable components in pages. Any updates, please?

<img width="584" alt="Screenshot 2024-02-15 at 11 37 04" src="https://github.com/vercel/next.js/assets/157653839/93e8eafc-672b-4b0f-b7a1-256f9955eedf">

**Comment by thexpand** [comment]
I have the exactly same issue as @AnastasiiaRoch. Using SCSS Modules. And it's not consistent, so it's sometimes difficult to reproduce. The problem is that it happens on production as well.

I think that the following issue might be related, but it says it only happens on dev mode, which is not the case: https://github.com/vercel/next.js/issues/13092

**Comment by scottsmith-1** [comment]
Using CSS Modules with and without compose and I've had all the same symptoms during route navigations.  Duplicates lots of styles and throws off the cascade completely.

**Comment by monolithed** [comment]
I'm just curious why the project team hasn't provided any comments on such a critical issue for several years and hasn't prioritized this task in any way.

**Comment by tinleym** [comment]
@ztanner does this pr address this issue? [https://github.com/vercel/next.js/pull/61428](url)

**Comment by IhorKevin** [comment]
It's another major version (14) and this pain still exists. I'm wondering why the issue is ignored by maintainers? Just look how much of CSS rules are useless because of the same component from project's UI kit is imported to different chunks:
![Screenshot 2024-03-05 at 12 05 30](https://github.com/vercel/next.js/assets/11776746/cc64b14a-d833-4d82-a9a0-90d76b817657)

**Comment by samcx** [comment]
We landed a <picture data-single-emoji=":pr:" title=":pr:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/pr/876af7e7d65e1595.png" alt=":pr:" align="absmiddle"></picture> → https://github.com/vercel/next.js/pull/63157.

Can anybody verify if this has been fixed on [v14.2.0-canary.28](https://github.com/vercel/next.js/releases/tag/v14.2.0-canary.28) or later?

**Comment by otaviosoares** [comment]
I need to run a few more tests, but it looks like it's fixed. 😃

**Comment by damnsamn** [comment]
Absolutely huge news, great work guys. Thank you!

**Comment by klonwar** [comment]
This is awesome! Looking forward to the release

**Comment by sannajammeh** [comment]
@samcx This is still an issue for us even on the latest Canary

<img width="293" alt="image" src="https://github.com/vercel/next.js/assets/50969683/ab6574e4-9572-48f4-8c52-a66fa60c692a">

**Comment by damnsamn** [comment]
> @samcx This is still an issue for us even on the latest Canary. Only happning on build.

Your screenshot doesn’t look like it depicts this issue, which is related to duplicate styles which override specificity due to how they’re ordered. This fix as I understand it corrects the CSS chunking algorithm to prevent importing the same component styles more than once.

**Comment by samcx** [comment]
@sannajammeh Can you share with us a simple <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> showcasing your issue and a explanation in the <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> (in a `README.md`) what we're supposed to see?

**Comment by IhorKevin** [comment]
I can not switch to canary and verify the issue is fixed (in my project I use Lingui SWC plugin which has an incompatibility with @swc/core used in canary branch of Next). But it's a big thing if it works. @samcx , thanks!

**Comment by codefln** [comment]
> We landed a <img alt=":pr:" width="20" height="auto" src="https://camo.githubusercontent.com/4fb083ea0a04edce4d651461f5913f28d39509d5d94d507458a0e7a10ee7cd23/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f70722f383736616637653764363565313539352e706e67"> → #63157.
> 
> Can anybody verify if this has been fixed on [v14.2.0-canary.28](https://github.com/vercel/next.js/releases/tag/v14.2.0-canary.28) or later?

I don't see duplicates anymore. When will the version not be in the canary?

**Comment by Stan-Stani** [comment]
I don't see duplicates anymore either! Nice job.

**Comment by rikers** [comment]
Same here, no more duplicates (dev server: OK / production build: OK).
Thanks!

**Comment by bhargav-spinx** [comment]
@rikers  @Stan-Stani @codefln @otaviosoares @damnsamn @klonwar 
How can we resolve this issue, and which version should we use to fix the duplication problem?
Thanks!

**Comment by codefln** [comment]
@bhargav-spinx

Sorry for the long answer. Just noticed your answer.
In version 14.2.0-canary.39 the bug has almost disappeared, but this is still the case if you use dynamic components.

Workaround:
1. Create a <Button /> component and a style file button.module.scss
2. Import the <Button /> component on the main page.
3. Create the <TestComponent /> component. import <Button /> in <TestComponent />
4. On another page, for example /about, we import the component dynamically: const DynamicTest = dynamic(() => import('../../TestComponent'));

We get a duplicate of styles.
Other times, I didn't notice it.

![telegram-cloud-photo-size-2-5215646192652900779-y](https://github.com/vercel/next.js/assets/24500468/7823404a-792d-4864-8630-e69fe08fdb8d)

**Comment by bhargavladani** [comment]
Fixed issues related to multiple instances, duplication, and out-of-order occurrences on Next.js version 14.

Modified the following files:
- `pages/_app.js`
- `pages/index.js`
- `styles/pages/_home.scss`
- `style/main.scss`

**pages/_app.js**
[Uploading Screen Recording 4-3-2024 at 10.36 AM.webm…]()

import Header from '@/app/components/Header';
import Footer from '@/app/components/Footer';

function MyApp({ Component, pageProps }) {
    return (
        `<Component {...pageProps} />`
    )
}
export default MyApp;

**pages/index.js**

import React from 'react'
import "../styles/sass/pages/_home.scss"

function index() {
    return (
        <>`<h1>home</h1>`</>
    )
}

export default index


**style/main.scss**

// Bootstrap Default components
@import "~bootstrap/scss/functions";
@import "~bootstrap/scss/variables";
@import "~bootstrap/scss/variables-dark";
@import "~bootstrap/scss/mixins";

// Project's custom variables or mixin if required
@import "sections/fonts";
@import "variables";
@import "functions";
@import "mixins";
@import "custom";

// Project core common styles
@import "layout";

Changes made:

1. `pages/_app.js`:
   - Corrected imports for `Header` and `Footer` components to use relative paths (`@/app/components/Header` and `@/app/components/Footer` respectively).

2. `pages/index.js`:
   - Imported the homepage component with the correct path (`"../styles/sass/pages/_home.scss"`) for styling.

3. `styles/pages/_home.scss`:
   - Ensured proper imports of `main.scss` and `reactSelect` in the homepage styles.
   - Updated styles for the `.banner` section to maintain consistency and compatibility with other components.

4. `style/main.scss`:
   - Checked compatibility with Next.js 14 by verifying imports and usage of Bootstrap and project-specific styles.
   - Verified that custom variables, mixins, and common styles are appropriately imported and utilized.

These changes aim to address compatibility issues with Next.js 14 and ensure consistency and correctness across the project files.

**Comment by terrymun** [comment]
@bhargavladani What you have posted is completely irrelevant to the issue at hand.

**Comment by bhargavladani** [comment]
@terrymun please check the video. i have fixed duplication css issues on my project.

**Comment by codefln** [comment]
@bhargavladani I don't know if you've read the thread, but there's a discussion here about style duplication issues with app-router.

In pages router, this bug never happened

**Comment by bhargavladani** [comment]
@codefln i think this thread is relevant to the issue below, right?"
https://github.com/vercel/next.js/issues/51030#issuecomment-2027204864

**Comment by mysticflute** [comment]
Is this fix for this also supposed to address the bug where imported css files in the root layout are not ordered as specified in the source?

https://github.com/vercel/next.js/issues/48120
https://github.com/vercel/next.js/issues/51213

**Comment by okonet** [comment]
I just upgraded to 14.2.2 and still experience the issue with vanilla-extract:

<img width="1208" alt="CleanShot 2024-04-19 at 11 16 44@2x" src="https://github.com/vercel/next.js/assets/11071/7a7672e3-628e-4d08-8b41-f010c0bc70f8">

**Comment by Christopher-md** [comment]
The same mistake. In version `v14.2.0-canary.39` everything is fine. But after upgrading to `v14.2.2`, duplicates appear again 😔

**Comment by okonet** [comment]
@samcx i think that sentiment is wrong. I still have the issue in development.

**Comment by samcx** [comment]
@okonet Can you provide a <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> so we can take a look?

**Comment by Christopher-md** [comment]
@samcx Here's how I managed to [reproduce](https://github.com/Christopher-md/next-css-bug) the error

Hope this helps

**Comment by hrigner** [comment]
I was struggling with this issue as well, and found a solution where all css is in one single file. While it might not be optimal, at least the styles are applied correct. This is what I added to the webpack config in the next config
```
config.optimization = {
      ...config.optimization,
      splitChunks: {
        cacheGroups: {
          styles: {
            name: "styles",
            type: "css/mini-extract",
            chunks: "all",
            enforce: true,
          },
        },
      },
    };
```

**Comment by pacocoursey** [comment]
Seeing the same issue with out-of-order CSS bundling from the root layout. Usually not a problem, but a big problem while trying to use cascade layers which rely on the browser parsing them in a specific order to determine their specificity.

Repro repo:
https://github.com/pacocoursey/next-cascade-layers

**Comment by samcx** [comment]
@pacocoursey Thanks for sharing the <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture>! Is this happening in both dev and deploy and or just deploy?

**Comment by mysticflute** [comment]
I updated to the latest and am still seeing this issue. Whenever the layout file uses a component that has `use client`, all of the global css in the layout file is ordered randomly, instead of in the order specified in the layout file. Happens in both dev and prod.

**Comment by brianmorin-ef** [comment]
To help with this issue, we're inserting our layer order onto every css file so the import order won't affect layer via next config and the BannerPlugin:

```
import webpack from "webpack";

const nextConfig = {
    webpack: (config) => {
        config.plugins.push(
            new webpack.BannerPlugin({
                banner: '@layer reset, ui, components;',
                test: /\.css$/,
                raw: true,
                entryOnly: false,
            })
        );

        return config;
    },
```

**Comment by terrymun** [comment]
I also read up about using `:where()` selector (that has zero specificity) as a possible alternative to using CSS layers, especially when working with component libraries that you have control over, but I haven't tested that out at the time of writing.

**Comment by kliatskiidenis** [comment]
I also have problems with duplicating styles when using scss modules. After making the build, the styles are still duplicated, I don't know what to do, because how will all this behave when the site is in production.
Please let me know if this problem will be solved and if it affects the css size?
<img width="1439" alt="Знімок екрана 2024-05-02 о 14 11 36" src="https://github.com/vercel/next.js/assets/128329970/1d8a43c2-fb39-4296-9914-bc7abc882377">

**Comment by codeaid** [comment]
I just spotted an issue where the order of my global imports is screwed up as soon as I use `Link` in one of my components instead of `a`, which doesn't make any sense.

I have the following two lines in my main layout:

```css
import 'normalize.css';
import './global.css';
```

The latter file overrides some of the heading styles and this is what I get without using `Link`:

<img width="375" alt="image" src="https://github.com/vercel/next.js/assets/637799/ef846ac2-4012-44da-a845-85564005966a">

and after I switch the `a` usage to `Link`:

<img width="375" alt="image" src="https://github.com/vercel/next.js/assets/637799/791f67b7-f779-44b3-b1eb-8f8fbb8e4728">

**Comment by Netail** [comment]
> I was struggling with this issue as well, and found a solution where all css is in one single file. While it might not be optimal, at least the styles are applied correct. This is what I added to the webpack config in the next config
> 
> ```
> config.optimization = {
>       ...config.optimization,
>       splitChunks: {
>         cacheGroups: {
>           styles: {
>             name: "styles",
>             type: "css/mini-extract",
>             chunks: "all",
>             enforce: true,
>           },
>         },
>       },
>     };
> ```

Does not seem to work great, when switching between pages, it's missing styles from the css bundle. Only once the page has been rendered at least once

**Comment by samcx** [comment]
Hi everyone!

We landed some fixes with [v15.0.0-canary.10](https://github.com/vercel/next.js/releases/tag/v15.0.0-canary.10).

Taking a look at some <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture>s now if I can still replicate CSS issues!

- https://github.com/pacocoursey/next-cascade-layers :fixed: 
-

**Comment by Netail** [comment]
Hi @samcx ,
I've tested the canary version right after publishing, but it seems like [my issue](https://github.com/vercel/next.js/issues/65480) with the following test setup;  https://github.com/Netail/app-dir-css-order was still facing duplicate css / incorrect css order  :(

**Comment by samcx** [comment]
@Netail Yeah can confirm we're still seeing issues—I've already let the team know and we're digging (<picture data-single-emoji=":dig-deeper:" title=":dig-deeper:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/dig-deeper/d3645e26292e183f.png" alt=":dig-deeper:" align="absmiddle"></picture>) <picture data-single-emoji=":frog-thumbsup:" title=":frog-thumbsup:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/frog-thumbsup/74768cbdae647b4d.png" alt=":frog-thumbsup:" align="absmiddle"></picture>

**Comment by Netail** [comment]
@samcx do you happen to know if it's possible to now fallback on non-streaming CSS (the original way)? Where it would send a single CSS file per path

**Comment by samcx** [comment]
@Netail Not 100% sure if this will help you, but you can try. →

```
experimental: {
  cssChunking: 'strict',
 }
 ```
 
 x-ref: https://github.com/vercel/next.js/pull/63157

**Comment by Netail** [comment]
@samcx Sadly not, feels like it makes it even worse if I look at the css order. So basically a server component uses a certain component styling, with some overwriting, these get added to the bundle. Then a client component renders which also uses the component styling, so this styling then gets added at the bottom (while it was already in the bundle), thus overwriting the overwrited component styling to the original styling.

**Comment by Netail** [comment]
Maybe I have some time tomorrow to help create some test cases in here, which can help you guys when fixing the issue?

**Comment by Netail** [comment]
But was more looking for a fallback solution (till this get sorted) where it would bundle the full page css in 1 css file, like the old school days.

**Comment by Netail** [comment]
Hi @samcx, I was wondering if there was any progress made internally the past 2 weeks 😅

**Comment by joshunger** [comment]
@samcx Thank you! When does `cssChunking` come out of experimental and get documentation around when to use it?

**Comment by kotAPI** [comment]
I am a maintainer on an open-source UI library, here's the docs website that runs on next 14 and is having this issue that's been discussed since over a year, we've no idea how to fix this

https://github.com/rad-ui/website

**Comment by Netail** [comment]
> I am a maintainer on an open-source UI library, here's the docs website that runs on next 14 and is having this issue that's been discussed since over a year, we've no idea how to fix this
> 
> https://github.com/rad-ui/website

Yeah this issue has been going on since v13 already :(

**Comment by ssijak** [comment]
I don't understand how this issue is not prioritised. It breaks so many projects in unpredictable ways, basically everything that does not import one CSS file like Tailwind.

**Comment by kutsan** [comment]
Sorry to ping everyone, but @shuding, @leerob, do you have any updates on this? It is affecting a lot of projects, and I also have questions about why it still hasn't been resolved.

**Comment by khanakia** [comment]
I have `import styles from './Spinner.module.scss'` and this CSS is loading two times like this

These both files contains the same spinner code I just can't figure why is it like this
```
layout.css
page.css
```

**Comment by samcx** [comment]
@kutsan The team was recently had an offsite, and then I was PTO for a week—we're back to tackling this issue! Sorry for the lack of an update.

**Comment by terrymun** [comment]
A report back from the last time I commented: I am working on a greenfield project and therefore has the ability to put in some mandates early on, and it seems that implementing CSS `@layer` resolve 99% of the issues we experienced.

This still only fixes the symptoms of having duplicate CSS being loaded, but now that all our components have mandatory `@layer` at-rule requirement (enforced with an in-house stylelint rule), we no longer experience issues with styles overriding each other.

Our next.js config looks like this:

```js
/**
 * @type {import('@nx/next/plugins/with-nx').WithNxOptions}
 */
const nextConfig = {
  ...restOfTheConfig,
  webpack: (config) => {
    config.plugins = [
      ...(config.plugins ?? []),
      new webpack.BannerPlugin({
        banner: `@layer ${CSS_LAYER_NAMES.join(', ')};`,
        test: /\.s?css$/,
        raw: true,
        entryOnly: false,
      }),
    ];

    return config;
  },
};
```

And `CSS_LAYER_NAME` is simply an array of CSS layers we have defined somewhere in a constant, as it is shared by our developer tooling and docs generation, but it looks like this:

```tsx
export const CSS_LAYER_NAMES = [
  'reset',
  'headless',
  'design-system',
  'component',
  'page',
  'layout',
  'app',
] as const;
```

**Comment by MONTYYYYY** [comment]
> @Netail Not 100% sure if this will help you, but you can try this in your `next.config.js`. →
> 
> ```
> experimental: {
>   cssChunking: 'strict',
> }
> ```
> 
> x-ref: #63157

This worked for me !!! Many thanks!

**Comment by samcx** [comment]
Ok back with a report!

- @Christopher-md Do you know which exact CSS issue there is with your <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> → https://github.com/Christopher-md/next-css-bug? There a several issues going around, so if that could be made clear, that would be very helpful.

- https://github.com/pacocoursey/next-cascade-layers → For the longest time, I didn't realize (my bad) that the _import order matters_ for CSS. So there was no issue with this <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture>. Just switch the `global.css` file up.

```
import * as React from "react";
import "./global.css"; // switched this to be above the Component, so the global css comes first
import { Component } from "./Component";

export default function Layout(props: React.PropsWithChildren) {
  return (
    <html lang="en">
      <body>
        {props.children}
        <Component />
      </body>
    </html>
  );
}
```

- @Netail Taking a look at this still → https://github.com/vercel/next.js/issues/65480 (I believe this is what you're referring to)

This <picture data-single-emoji=":pr:" title=":pr:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/pr/876af7e7d65e1595.png" alt=":pr:" align="absmiddle"></picture> https://github.com/vercel/next.js/pull/67373 is going to land as well, which fixes https://github.com/vercel/next.js/issues/64773. This may fix some of the other <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> I may have missed.

**Comment by Netail** [comment]
> (I believe this is what you're referring to)

Yes :) Might be related to one of the other reported issues tho, but not 100% sure

**Comment by kotAPI** [comment]
> cssChunking: 'strict',

On Next 14 🫠 

```
Invalid next.config.js options detected: 
 ⚠     Unrecognized key(s) in object: 'cssChunking' at "experimental"
 ```

**Comment by klonwar** [comment]
> For the longest time, I didn't realize (my bad) that the import order matters for CSS. So there was no issue with this :repro:. Just switch the global.css file up.

@samcx Man, you're my hero, thank you! Eslint [simple-import-sort](https://www.npmjs.com/package/eslint-plugin-simple-import-sort) plugin set me up 🥲 Disabled it in layout.tsx with `// eslint-disable-next-line simple-import-sort/imports`

**Comment by dfd1123** [comment]
There is no problem when building, but the problem occurs only when using dev. When can it be fixed? This is making it very difficult for me.

> 좋아, 보고서를 다시 제출하겠습니다!
> 
> * @Christopher-md귀하의 CSS에 어떤 정확한 문제가 있는지 알고 계십니까?<img alt=":재생산:" width="20" height="auto" src="https://camo.githubusercontent.com/294fcc016906904beb80b3c14e2252d0b6fef2711a1acb2afaac1c89b8427edc/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f726570726f2f356365613533623535333238323562342e706e67">→ https://github.com/Christopher-md/next-css-bug ? 몇 가지 문제가 돌고 있으니, 그것을 명확히 밝혀 주시면 매우 도움이 될 것입니다.
> * https://github.com/pacocoursey/next-cascade-layers → 아주 오랫동안, 저는 CSS에 대한 _가져오기 순서가 중요하다는_ 것을 깨닫지 못했습니다(제 잘못입니다) . 그래서 이것에는 문제가 없었습니다.<img alt=":재생산:" width="20" height="auto" src="https://camo.githubusercontent.com/294fcc016906904beb80b3c14e2252d0b6fef2711a1acb2afaac1c89b8427edc/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f726570726f2f356365613533623535333238323562342e706e67">. `global.css`파일을 바꾸면 됩니다.
> 
> ```
> import * as React from "react";
> import "./global.css"; // switched this to be above the Component, so the global css comes first
> import { Component } from "./Component";
> 
> export default function Layout(props: React.PropsWithChildren) {
>   return (
>     <html lang="en">
>       <body>
>         {props.children}
>         <Component />
>       </body>
>     </html>
>   );
> }
> ```
> 
> * @Netail이 스틸을 살펴보세요 → [서버와 클라이언트 구성 요소의 조합으로 인해 CSS 순서가 잘못되었습니다 ](https://github.com/vercel/next.js/issues/65480)[ #65480](https://github.com/vercel/next.js/issues/65480) (이것이 당신이 언급한 내용인 것 같습니다)[](https://github.com/vercel/next.js/issues/65480)
> 
> 이것<img alt=":홍보:" width="20" height="auto" src="https://camo.githubusercontent.com/4fb083ea0a04edce4d651461f5913f28d39509d5d94d507458a0e7a10ee7cd23/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f70722f383736616637653764363565313539352e706e67"> #67373 도 착륙할 예정이며, 이는 [#64773을](https://github.com/vercel/next.js/issues/64773) 수정합니다 . 이는 다른 일부 문제를 수정할 수 있습니다.<img alt=":재생산:" width="20" height="auto" src="https://camo.githubusercontent.com/294fcc016906904beb80b3c14e2252d0b6fef2711a1acb2afaac1c89b8427edc/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f726570726f2f356365613533623535333238323562342e706e67">내가 놓친 게 있을 수도 있어요.
> 
> 지금까지는 확인된 것으로 적극적으로 알려진 시급한 이슈들은 다음과 같다고 생각합니다.<img alt=":재생산:" width="20" height="auto" src="https://camo.githubusercontent.com/294fcc016906904beb80b3c14e2252d0b6fef2711a1acb2afaac1c89b8427edc/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f726570726f2f356365613533623535333238323562342e706e67">s는 다음과 같습니다:
> 
> 1. 서버 ↔ 클라이언트 구성 요소가 CSS 가져오기 순서를 따르지 않습니다.
> 2. CSS는 경로 그룹의 레이아웃 사이에서 누출됩니다.<img alt=":결정된:" width="20" height="auto" src="https://camo.githubusercontent.com/f853dd142707234ff7e973151c99f6fda94ab821f85398fe02a1963269a82e9c/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f66697865642f363662623834353566323462643530652e706e67">[다른 청크 그룹으로 누출되는 방식으로 글로벌 CSS를 병합하지 않도록 합니다](https://github.com/vercel/next.js/pull/67373) ( [ #67373](https://github.com/vercel/next.js/pull/67373) ). 병합을 피하면 `global.css`다른 문제가 해결될지 100% 확신할 수는 없지만, 살펴보도록 하겠습니다!

There is no problem when building, but the problem occurs only when using dev. When can it be fixed? This is making it very difficult for me.

**Comment by samcx** [comment]
@kotAPI Your previous post's reply implies you're not on a later version with the first fix for ordering/merging—that's why you're not able to use `cssChunking: 'strict'`. Are you able to upgrade to the latest 14 canary or the the current latest canary version?

**Comment by khanakia** [comment]
@PASHANSX I also had the same issue I mentioned above what I did was move the CSS code out of the module and put it inside global.css

This may be how next.js 14 is designed which generated different CSS files for layout and page CSS; there is nothing you can do about this.

**Comment by aniketpadmansh** [comment]
<img width="588" alt="Screenshot 2024-07-05 at 10 30 01 PM" src="https://github.com/vercel/next.js/assets/171565780/1cba5cd7-4b1d-4868-83c8-39b20b4c71c6">
un necessary css is being bundled with each page

**Comment by Netail** [comment]
> <img width="588" alt="Screenshot 2024-07-05 at 10 30 01 PM" src="https://github.com/vercel/next.js/assets/171565780/1cba5cd7-4b1d-4868-83c8-39b20b4c71c6">
> 
> un necessary css is being bundled with each page
> 

Those are chunks no? It basically fetches css on demand

**Comment by aniketpadmansh** [comment]
I am using module.scss
so scss writtn in b and c and not used on page a is also getting downloaded for page a

**Comment by samcx** [comment]
@kotAPI You need to be at least on [v14.2.0-canary.28](https://github.com/vercel/next.js/releases/tag/v14.2.0-canary.28)—14.1 would not have that this <picture data-single-emoji=":pr:" title=":pr:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/pr/876af7e7d65e1595.png" alt=":pr:" align="absmiddle"></picture> → https://github.com/vercel/next.js/pull/63157.

**Comment by Christopher-md** [comment]
> Ok back with a report!
> 
> * @Christopher-md Do you know which exact CSS issue there is with your <img alt=":repro:" width="20" height="auto" src="https://camo.githubusercontent.com/294fcc016906904beb80b3c14e2252d0b6fef2711a1acb2afaac1c89b8427edc/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f726570726f2f356365613533623535333238323562342e706e67"> → https://github.com/Christopher-md/next-css-bug? There a several issues going around, so if that could be made clear, that would be very helpful.
> * https://github.com/pacocoursey/next-cascade-layers → For the longest time, I didn't realize (my bad) that the _import order matters_ for CSS. So there was no issue with this <img alt=":repro:" width="20" height="auto" src="https://camo.githubusercontent.com/294fcc016906904beb80b3c14e2252d0b6fef2711a1acb2afaac1c89b8427edc/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f726570726f2f356365613533623535333238323562342e706e67">. Just switch the `global.css` file up.
> 
> ```
> import * as React from "react";
> import "./global.css"; // switched this to be above the Component, so the global css comes first
> import { Component } from "./Component";
> 
> export default function Layout(props: React.PropsWithChildren) {
>   return (
>     <html lang="en">
>       <body>
>         {props.children}
>         <Component />
>       </body>
>     </html>
>   );
> }
> ```
> 
> * @Netail Taking a look at this still → [CSS order incorrect with the combination of server and client components #65480](https://github.com/vercel/next.js/issues/65480) (I believe this is what you're referring to)
> 
> This <img alt=":pr:" width="20" height="auto" src="https://camo.githubusercontent.com/4fb083ea0a04edce4d651461f5913f28d39509d5d94d507458a0e7a10ee7cd23/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f70722f383736616637653764363565313539352e706e67"> #67373 is going to land as well, which fixes #64773. This may fix some of the other <img alt=":repro:" width="20" height="auto" src="https://camo.githubusercontent.com/294fcc016906904beb80b3c14e2252d0b6fef2711a1acb2afaac1c89b8427edc/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f726570726f2f356365613533623535333238323562342e706e67"> I may have missed.
> 
> So far, I think the pressing issues that are actively known with confirmed <img alt=":repro:" width="20" height="auto" src="https://camo.githubusercontent.com/294fcc016906904beb80b3c14e2252d0b6fef2711a1acb2afaac1c89b8427edc/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f726570726f2f356365613533623535333238323562342e706e67">s are:
> 
> 1. Server ↔ Client components are not respecting CSS import order
> 2. CSS is leaking between Layouts in Route Groups (getting <img alt=":fixed:" width="20" height="auto" src="https://camo.githubusercontent.com/f853dd142707234ff7e973151c99f6fda94ab821f85398fe02a1963269a82e9c/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f66697865642f363662623834353566323462643530652e706e67"> by [avoid merging global css in a way that leaks into other chunk groups #67373](https://github.com/vercel/next.js/pull/67373)). Not 100% sure if avoiding merging the `global.css` will fix other issues, but will be taking a look!

@samcx, Hi

The problem arose after the implementation of the functionality for opening modal windows.

The [modal window](https://codesandbox.io/p/github/Christopher-md/next-css-bug/main?file=%2Fsrc%2Flayouts%2FMainLayout%2F_components%2FHeader%2FHeader.tsx) contains the Input component, which is also used on the page. 

As a result, there is a duplicate of styles for Input

Pay attention to the [functionality](https://codesandbox.io/p/github/Christopher-md/next-css-bug/main?file=%2Fsrc%2Fcontexts%2FModalContext%2FModalContext.tsx) of opening these modal windows and how it is used

Glad to help

**Comment by AndrePessoa** [comment]
@samcx, first, thank you for your amazing work! 

Second, the experiment flag worked for me on my team project and it seems to be helping more people. Do you plan to release it as a patch any time soon?

**Comment by Netail** [comment]
> Server ↔ Client components are not respecting CSS import order

Actually is seems more like the following;

If the layout uses the test component and passes a override class to the object, but a page also uses the test component.
The style sheets for both will have the test component's css imported with the same class names, but as the page scss is injected into the html after the layout scss, the default css class will be used last, thus resetting the override css

<img width="901" alt="Screenshot 2024-07-10 at 15 22 40" src="https://github.com/vercel/next.js/assets/11695769/dac1979c-6f8e-4ef1-b4ec-bae779551fda">

<img width="335" alt="Screenshot 2024-07-10 at 15 22 15" src="https://github.com/vercel/next.js/assets/11695769/3f7c2dd2-9781-478a-8308-bb91ec488b7c">

**Comment by samcx** [comment]
> Updated the Next version.js is up to the latest version and noticed that duplicates occur if the component is used in both page.tsx and layout.tsx; If a component is used in multiple layout.tsx, then the number of duplicates corresponds to the number of uses in layout.tsx

@Christopher-md Can you try minimizing your <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture>? I don't see examples for either cases.

@AndrePessoa No plan yet. I think ideally, this shouldn't even be an option—we should be able to minimize the number of requests (merging properly) without breaking CSS. There was an additional [fix](https://github.com/vercel/next.js/pull/67373) (also just backported to [14.2.5](https://github.com/vercel/next.js/releases/tag/v14.2.5)), which may allow for not having to use the experimental option.

@Netail Is this in your <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> in your other reported issue?

**Comment by Netail** [comment]
> @Netail Is this showcased in your <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> in your other reported issue? <picture data-single-emoji=":frog-eyes:" title=":frog-eyes:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/frog-eyes/558eb9b0fd306c97.png" alt=":frog-eyes:" align="absmiddle"></picture> 
> 

Yes! That's the example used :)

**Comment by kutsan** [comment]
@samcx I'm still having the problem, even with the v14.2.5 release.

Here's my repro: https://github.com/kutsan/nextjs-css-duplicates-issue

When I inspect the title styles, I'm seeing duplicates.

So far, I've found two possible solutions:

1. In `@/app/layout.tsx`, exporting from `@/features/RootLayout` instead of `@/features` fixes it and removes the duplicate styles.
2. Removing `'use client'` in the `@/features/RootRoute.tsx` file also fixes it.

I'm not sure why these work, but I hope this helps.

<img width="1442" alt="image" src="https://github.com/vercel/next.js/assets/10108377/16a9ac55-dbf0-44d6-a190-ab00febcf1e3">

**Comment by Christopher-md** [comment]
> > Updated the Next version.js is up to the latest version and noticed that duplicates occur if the component is used in both page.tsx and layout.tsx; If a component is used in multiple layout.tsx, then the number of duplicates corresponds to the number of uses in layout.tsx
> 
> @Christopher-md Can you try minimizing your <img alt=":repro:" width="20" height="auto" src="https://camo.githubusercontent.com/294fcc016906904beb80b3c14e2252d0b6fef2711a1acb2afaac1c89b8427edc/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f726570726f2f356365613533623535333238323562342e706e67">? I don't see examples for either cases.
> 
> @AndrePessoa No plan yet. I think ideally, this shouldn't even be an option—we should be able to minimize the number of requests (merging properly) without breaking CSS. There was an additional [fix](https://github.com/vercel/next.js/pull/67373) (also just backported to [14.2.5](https://github.com/vercel/next.js/releases/tag/v14.2.5)), which may allow for not having to use the experimental option.
> 
> @Netail Is this showcased in your <img alt=":repro:" width="20" height="auto" src="https://camo.githubusercontent.com/294fcc016906904beb80b3c14e2252d0b6fef2711a1acb2afaac1c89b8427edc/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f726570726f2f356365613533623535333238323562342e706e67"> in your other reported issue? <img alt=":frog-eyes:" width="20" height="auto" src="https://camo.githubusercontent.com/30dc99a16c350c28c7116a2f03e12a6ef87cf51c1ca2e0db99a390a9e809ab3d/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f66726f672d657965732f353538656239623066643330366339372e706e67">

@samcx 

Here is a minimized [RE-PRO](https://codesandbox.io/p/github/Christopher-md/next-css-bug/main?file=%2Fsrc%2Fapp%2F%28main%29%2Fpage.tsx%3A9%2C2&workspaceId=43aaa7c7-9a54-4006-8889-cc20b9ceb408)

**Comment by Netail** [comment]
> I have not words.
> 
> Was developing my whole app for a month with `--turbo` relying on CSS cascades heavily just to found out **it's completely broken after the build**
> 
> Is this the DX improvements we were promised with Turbo (which no one asked for)?

Turbopack is ready for development, production still has 537 test remaining to be fixed (https://areweturboyet.com/build)

**Comment by khanakia** [comment]
I switched to Astro.js for my new project

**Comment by bruno12mota** [comment]
> > I have not words.
> > Was developing my whole app for a month with `--turbo` relying on CSS cascades heavily just to found out **it's completely broken after the build**
> > Is this the DX improvements we were promised with Turbo (which no one asked for)?
> 
> Turbopack is only ready for development, production still has 537 test remaining to be fixed (https://areweturboyet.com/build). Most of the remaining tests are related to CSS

It is happening in turbo development tough, I've opened a new issue https://github.com/vercel/next.js/issues/68412 since it might not be entirely related with this one and has a minimal repo to showcase the issue.

**Comment by pixelchutes** [comment]
Following. I believe I am also running into this. Even with the merge of #67373, when using `--turbo` on `v14.2.5`, I am seeing CSS Module specific issues, primarily local dev UI is not matching production.

Further, HMR is throwing console errors for any changes within `.module.css`:

> `Error: No link element found for chunk static/chunks/_793f98._.css`

<img width="483" alt="image" src="https://github.com/user-attachments/assets/9ecc21fb-c97d-4073-b1dc-ab89bb520524">

**Comment by samcx** [comment]
@pixelchutes Is it possible to create a minimal <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> so we can take closer a look at your case?

**Comment by pixelchutes** [comment]
@samcx Let me see if I can put something together! Creating a reminder to revisit early next week 👍🏼

**Comment by subproject22** [comment]
@samcx I think Paco made a perfect repro of this issue before, it was ignored for some reason. Here's is it, did you saw it? https://github.com/pacocoursey/next-cascade-layers

**Comment by Netail** [comment]
Are there any internally discussed updates we should know of? CSS is a pretty crucial part 😥

**Comment by samcx** [comment]
@subproject22 Paco's example wasn't ignored. There was actually no broken behavior here, we just had to update the import order, which _matters_. I shared this here (earlier comment) → https://github.com/vercel/next.js/issues/51030#issuecomment-2203871536.

@Netail Nothing to share yet. Will be considering moving this to discussions soon, as this thread has becoming too broad (everyone is just piling their CSS issue here)—we need to actually isolate an issue to the actual issue at play, such as your Client ↔ Server CSS issue that you reported.

We are also tackling this old CSS issue as well → https://github.com/vercel/next.js/issues/33286.

**Comment by Enkratia** [comment]
@samcx i created default next.js app, created 1 server component with 1 LIink and 1 Image, 1 css module. Can you explain why HMR doesn't work? This problem persists more than 1 year.

If repro is needed: https://github.com/Enkratia/css-reload-show

**Comment by samcx** [comment]
@Enkratia Please open a separate [bug report](https://github.com/vercel/next.js/issues/new?assignees=&labels=template:+bug&projects=&template=1.bug_report.yml) detailing your HMR issue with CSS.
